### PR TITLE
forwardport PR#110

### PR DESCRIPTION
**Pull request by Tyrael**
this PR is basically the same as #110 but it should apply to the current master without any conflicts.

### CODE REVIEW

**Comment by samifruit514** [comment]
I don't get why this change is not merged yet. It is very useful when we push new tags into a git repo.

**Comment by mbrodala** [comment]
@Seldaek What's your opinion on this one?

Given the fact that webhooks from Github/Bitbucket only contain Git repository URLs and not Composer package names, this change would be very useful to trigger updates for single packages.

**Comment by naderman** [comment]
Finally got around to look at this. So first thing that comes to mind is: Why doesn't this use the merging of packages as the package filter does, see https://github.com/Tyrael/satis/blob/partial_update/src/Composer/Satis/Command/BuildCommand.php#L195 - is the cache file really needed?

**Comment by Tyrael** [comment]
#110 was created before the packages filter was added (originally with 4e9780a980f3270eda98ba9a5797cce9463974f0 then later renamed)

**Comment by naderman** [comment]
@Tyrael I realize the history, but would be nice to update this now then if possible?

**Comment by Tyrael** [comment]
sure, and I will update the PR, I was just answering your question why was this not done that way in the first place :)

**Comment by Tyrael** [comment]
@naderman any chance you had the time to review this?

File: src/Composer/Satis/Command/BuildCommand.php [hunk 1]
```diff
@@ -145,6 +146,20 @@ protected function execute(InputInterface $input, OutputInterface $output)
             throw new \InvalidArgumentException('The output dir must be specified as second argument or be configured inside '.$input->getArgument('file'));
         }
 
+        if (($singleRepositoryUrl = $input->getOption('repository-url')) !== null) {
+            $singleRepositoryConfig = null;
+            foreach ($config['repositories'] as $r) {
+                if ($r['url'] == $singleRepositoryUrl) {
```
Comment (stof) [inline src/Composer/Satis/Command/BuildCommand.php]: should use strict comparison

File: src/Composer/Satis/Command/BuildCommand.php [hunk 2]
```diff
@@ -145,6 +146,20 @@ protected function execute(InputInterface $input, OutputInterface $output)
             throw new \InvalidArgumentException('The output dir must be specified as second argument or be configured inside '.$input->getArgument('file'));
         }
 
+        if (($singleRepositoryUrl = $input->getOption('repository-url')) !== null) {
+            $singleRepositoryConfig = null;
+            foreach ($config['repositories'] as $r) {
```
Comment (stof) [inline src/Composer/Satis/Command/BuildCommand.php]: you cannot assume that this array key exists. (nor that ``url`` exists in it) as ``$config`` is just the json-decoded content of the config file, and so you have no idea about what it contains or what is not in it.

**Comment by Tyrael** [comment]
while going through the comments from @stof I've just realized that @naderman suggestion in https://github.com/composer/satis/pull/232#issuecomment-138104593 was probably that we should instead extend selectPackages() and loadDumpedPackages() with a $repositoriesFilter similarly how it was done for $packagesFilter.
@naderman if that's what you wanted just let me now and I can do that.

**Comment by realshadow** [comment]
How does it look? I could really use this, I have patched satis for now

**Comment by alcohol** [comment]
It needs to be rebased unfortunately.

**Comment by realshadow** [comment]
@Tyrael Can you please try to move this forward if you have time?

Or I can try it if you are busy.

**Comment by Tyrael** [comment]
originally I just forwardported the quick and dirty solution but then I realized/was suggested a better approach but then got swamped with work at my dayjob and couldn't finish it.
I'm not given up on this but I don't think I can work on it for like another 2-3 weeks.

**Comment by realshadow** [comment]
Just for clarification...

I assume combining `--repository-url` and `package` in one command is reduntant (and currently breaks packages). Should they be handled as mutually exclusive? 

Wouldn't it be better/easier to reuse `package` for repository URL as well and then based on argument type (repository URL, package name) get correct repository? Even though that might be a bit confusing as far as arguments go.

**Comment by realshadow** [comment]
I looked at the latest code and it moved miles ahead...

After a bit of searching I found that this could be easily solved with just changing `findPackages` in `Composer\Repository\ArrayRepository` to include comparison of `$package->getSourceUrl()` (or create a separate method for that) and then it could be used together with package filter, e.g.

```
php bin\satis build satis.json public psr/log https://github.com/realshadow/serializers.git
```

Although I am not sure if it should go there. Any thoughts @alcohol, @Tyrael ?

**Comment by alcohol** [comment]
@realshadow you mean you want to introduce changes to Composer for a feature in Satis? That is unlikely to happen unless you have a really good reason. Also I kind of lost track of what this PR was trying to accomplish since it's a port of a port (and I don't use Satis myself).

**Comment by realshadow** [comment]
@alcohol Not really. Just saying :) 

It was trying to accomplish building of a single/multiple repositories with repository url instead of package name. The original proposal was for `--repository-url` argument. @Tyrael  then wanted to do  like suggested in https://github.com/composer/satis/pull/232#issuecomment-142918732, but since then    satis moved miles ahead and both mentioned methods do not exist anymore, instead the filtering is handled by composer itself via `findPackages` (thus my comment above).

Although on second thought, the original code, with slight alterations should be enough to cover this use case...

**Comment by alcohol** [comment]
The code was mostly broken into smaller manageable pieces. It should still exist to some extend.

**Comment by realshadow** [comment]
I have created a new PR https://github.com/composer/satis/pull/266 which uses `PackageSelection` for filtering.

**Comment by JamesRezo** [comment]
I guess it can be closed ;)

**Pull request closed without merging**